homeassistant/components/media_player/universal.py

```diff
@@ -0,0 +1,437 @@
+
+        if any([cmd in self._cmds for cmd in [SERVICE_VOLUME_UP,
+                                              SERVICE_VOLUME_DOWN]]):
+            flags |= SUPPORT_VOLUME_STEP
+            flags &= ~SUPPORT_VOLUME_SET
+
+        if SERVICE_VOLUME_MUTE in self._cmds and \
+                ATTR_MEDIA_VOLUME_MUTED in self._attrs:
+            flags |= SUPPORT_VOLUME_MUTE
+
+        return flags
+
+    @property
+    def device_state_attributes(self):
+        """ Extra attributes a device wants to expose. """
+        active_child = self.active_child_state
+        return {ATTR_ACTIVE_CHILD: active_child.entity_id} \
+            if active_child else {}
+
+    def turn_on(self):
+        """ turn the media player on. """
+        self._call_service(SERVICE_TURN_ON, allow_override=True)
+
+    def turn_off(self):
+        """ turn the media player off. """
+        self._call_service(SERVICE_TURN_OFF, allow_override=True)
+
+    def mute_volume(self, is_volume_muted):
+        """ mute the volume. """
+        data = {ATTR_MEDIA_VOLUME_MUTED: is_volume_muted}
+        self._call_service(SERVICE_VOLUME_MUTE, data, allow_override=True)
+
+    def set_volume_level(self, volume_level):
+        """ set volume level, range 0..1. """
+        data = {ATTR_MEDIA_VOLUME_LEVEL: volume_level}
+        self._call_service(SERVICE_VOLUME_SET, data)
+
+    def media_play(self):
+        """ Send play commmand. """
+        self._call_service(SERVICE_MEDIA_PLAY)
+
+    def media_pause(self):
+        """ Send pause command. """
+        self._call_service(SERVICE_MEDIA_PAUSE)
+
+    def media_previous_track(self):
+        """ Send previous track command. """
+        self._call_service(SERVICE_MEDIA_PREVIOUS_TRACK)
+
+    def media_next_track(self):
+        """ Send next track command. """
+        self._call_service(SERVICE_MEDIA_NEXT_TRACK)
+
+    def media_seek(self, position):
+        """ Send seek command. """
+        data = {ATTR_MEDIA_SEEK_POSITION: position}
+        self._call_service(SERVICE_MEDIA_SEEK, data)
+
+    def play_youtube(self, media_id):
+        """ Plays a YouTube media. """
+        data = {'media_id': media_id}
+        self._call_service(SERVICE_YOUTUBE_VIDEO, data)
+
+    def play_media(self, media_type, media_id):
+        """ Plays a piece of media. """
+        data = {'media_type': media_type, 'media_id': media_id}
+        self._call_service(SERVICE_PLAY_MEDIA, data)
+
+    def volume_up(self):
+        """ volume_up media player. """
+        self._call_service(SERVICE_VOLUME_UP, allow_override=True)
+
+    def volume_down(self):
+        """ volume_down media player. """
+        self._call_service(SERVICE_VOLUME_DOWN, allow_override=True)
+
+    def media_play_pause(self):
+        """ media_play_pause media player. """
+        self._call_service(SERVICE_MEDIA_PLAY_PAUSE)
+
+    def update_state(self, *_):
+        """ event to trigger a state update in HA """
+        self._cache_active_child_state()
```

Comment:
The problem is that this function is the one calling update_ha_state. So it would just sit in a loop.